src/inference/forward.jl

```diff
@@ -99,12 +100,15 @@ function _forward!(
         αₜ = view(α, :, t)
         Bₜ = view(B, :, t)
         if t == t1
-            copyto!(αₜ, initialization(hmm))
+            copyto!(αₜ, initialization(hmm, control_seq[t]))
         else
             αₜ₋₁ = view(α, :, t - 1)
-            predict_next_state!(αₜ, hmm, αₜ₋₁, control_seq[t - 1])
+            predict_next_state!(αₜ, hmm, αₜ₋₁, control_seq[t]) # If `t` influences the transition from time `t` to `t+1` (and not from time `t-1` to `t`), then the associated control must be at `t+1`, right? If `control_seq[t-1]`, then we're using the control associated with the previous state and not the correct control, aren't we? The transition matrix would be P(X_{t}|X_{t-1},U_{t-1}) and not P(X_{t}|X_{t-1},U_{t}) as it should be. E.g., if `t == t1 + 1`, then `αₜ₋₁ = view(α, :, t1)` and the function would use the transition matrix P(X_{t1+1}|X_{t1},U_{t1}) instead of P(X_{t1+1}|X_{t1},U_{t1+1}). Same at logdensity.jl, line 37; forward_backward.jl, line 53.
```

Comment:
Whether the transition `t-1 -> t` is influenced by control `u[t]` or `u[t-1]` is an arbitrary choice. It doesn't really matter as long as we apply it consistently. Indeed, the transition happens "between time steps" so there is no right or wrong pick.
My choice for this package is to use `u[t-1]` (the previous time step), as explained in the [`transition_matrix` docstring](https://gdalle.github.io/HiddenMarkovModels.jl/stable/api/#HiddenMarkovModels.transition_matrix). It is really important that you respect it, otherwise it will be a breaking and very confusing change for users.

Reply:
You're right, it is an arbitrary choice, and you can adapt it to use any other indexation, as long as it is consistent. People using 0-based indexing are a hard proof of this! However, there are consequences and some do affect the user and the expected behaviour. I'd argue the following:


Input/Output HMM is a particular case of a HMM with covariates. Usually, with the exception of autoregression, the standard behaviour is to assume that if $\mathbf{u} = (u_{1},\ldots,u_{N})$ is a covariate of $\mathbf{x} = (x_{1},\ldots,x_{N})$, then $x_{t} = f(u_{t}) + \varepsilon_{t}$. In the transition matrix, we have that $x_{t}$ depends on $x_{t-1}$, which is the same as saying that it follows an AR(1): $x_{t} = f(x_{t-1}) + \varepsilon_{t}$. With an additional covariate, then it is natural to assume $x_{t} = f(x_{t-1}, u_{t}) + \varepsilon_{t}$. I mean... I believe we can both agree that when give as an input a data table (or analogous structure) to do a regression analysis with a response column $y$ and covariates $x, z$, we would expect that the function would do $y_{t} = f(x_{t}, z_{t}) + \varepsilon_{t}$.


In the majority of the literature I've read, the control variable index matches the targeted time: $P(X_{t}|X_{t-1},U_{t})$ or $P(X_{t+1}|X_{t},U_{t+1})$. It is the case from [Bengio & Frasconi (1994)](https://proceedings.neurips.cc/paper_files/paper/1994/file/8065d07da4a77621450aa84fee5656d9-Paper.pdf) up to [Visser & Speekenbrink (2022)](https://link.springer.com/book/10.1007/978-3-031-01440-6), including [Langrock, MacDonald & Zucchini (2016)](https://www.taylorfrancis.com/books/mono/10.1201/b20790/hidden-markov-models-time-series-walter-zucchini-iain-macdonald-roland-langrock). There are exceptions, of course, but mostly when there is a specific model which incorporate the covariate only in the transition probability (x)or only in the emission probability.


To be fair, and make a counterpoint to my own position, Visser & Speekenbrink, which are the maintainers of [depmixS4](https://cran.r-project.org/web/packages/depmixS4/index.html), did something *very strange* in their article describing the package. On page 3 of their [article](https://cran.r-project.org/web//packages/depmixS4/vignettes/depmixS4.pdf) you can notice that the control vector $\mathbf{z} = (z_{1},\ldots,z_{N})$ has a component $z_{1}$ that affects both the initial distribution *and* the first state transition. But it is the only case in which a covariate $z_{t}$ affects both the current and the upcoming time! This directly contradicts what is written on page 163, equation (4.48), of their aforementioned book, in which - once again - the control variable index matches the targeted time. To add insult to injury, in the very next equation (4.49) for a specific example, they once again use $P(X_{t+1}|X_{t},U_{t})$, but in a way that $z_{N}$ is effectively useless.


In fact, if we decide to make it so that the covariate $\mathbf{z}$ only affects the initial and transition probabilities, and not the emission probabilities (as shown both in Langrock, MacDonald & Zucchini and Visser & Speekenbrink), then the last element $z_{N}$ would be useless. For parsimony, it should be excluded, but, if we do it, then the length of the state sequence won't be equal to the length of the control sequence. One element, $z_{1}$ affects both current and upcoming time, and $z_{N}$ does nothing. Is this expected behaviour?


The code as written makes the user determine how the control variable should affect the parameters via `transition_matrix` and `obs_distributions`, but was silent on ` initialization`. However, if we want the control variable to only affect the initial and the transition probabilities, then `u[1]` would affect both the initial probability and the first transition, while `u[N]` would do nothing. On the other hand, if we want the control variable to only affect the transition probabilities, then `u[N]` would do nothing. I get it, I understand that if we approach the control as being "between time steps", then, naturally, the length of the control sequence should be the length of the state sequence minus 1 (aka the number of transitions). But I'd say that this approach is too specific: it only works for transition probabilities, it isn't the case for either initial or emission probabilities, and you're doing here a framework that is very broad by design, don't you? Isn't it better to assume that all data the user is giving as an input is useful data and don't let some part of this information to be useless? And if the user wants to use a model with the control being "between time steps", it is simple as checking if the length of the control sequence is equal to the observation sequence and, if it is 1 element shorter, then prepend a `nothing` or something like this. For all other cases, doesn't it make at least a little sense?


Now, just to make sure: of course, you're the boss. If you are sure that this is the way, I will undo these changes and adapt within the frame you previously specified. I did this just to include the control variable in the initial distribution in a way that made sense to me, but I'm open to suggestions of how to do so in a different manner! (...and I will answer why I believe it is reasonable to add control to the initial distribution in the appropriate topic below).

Reply:
I have opened #125 to keep track of this discussion. If you say that the convention you propose is the one present in the literature, then it makes sense to switch. However:
- it is another change that is orthogonal to adding autoregression, so it should go in a separate PR (either before or after this one)
- it is breaking so it needs to be clearly documented
- pleaaaaaaase for the love of God write shorter comments on your PRs :rofl: